### PR TITLE
Attempt to make load `from_tf=True` more efficiently

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Remove the need to get 2X RAM when loading from h5 file
- Remove the need to have TF present to load `from_tf=True` (This is reduced to h5 only)

A much better version (imo) of the finale code resides here:

https://github.com/Narsil/transformers/pull/4

Keeping separate PRs since it's a more involved change.

Script:
```python
from transformers import GPT2Model

model = GPT2Model.from_pretrained("gpt2", from_tf=True)
```

Making sure network is not a thing:
```
TRANSFORMERS_OFFLINE=1 /usr/bin/time -v python test.py
```

Master:
```
	Command being timed: "python test.py"
	User time (seconds): 9.90
	System time (seconds): 2.24
	Percent of CPU this job got: 224%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:05.42
	Maximum resident set size (kbytes): 2411224
```

With this PR:
```
	Command being timed: "python test.py"
	User time (seconds): 2.41
	System time (seconds): 0.97
	Percent of CPU this job got: 161%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.09
	Maximum resident set size (kbytes): 1096128
```

And just a comparison point loading from Pickled file directly:
```
	Command being timed: "python test.py"
	User time (seconds): 2.63
	System time (seconds): 0.95
	Percent of CPU this job got: 178%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.00
	Maximum resident set size (kbytes): 1319596
```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by Rocketknight1 — COMMENTED** [review]
This looks like a really cool PR, to totally eliminate the dependency on TF when crossloading TF weights into PT! I suspect there will definitely be issues surrounding the naming of weight tensors that are probably special-cased in our loading code right now, but it should be possible to make this work with a few small edits.

**Comment by Narsil** [comment]
@sgugger @LysandreJik I think all the tests are passing now. Is there any thing I should be extra careful about with this / test in some form ?

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 1]
```diff
@@ -300,26 +410,110 @@ def load_tf2_checkpoint_in_pytorch_model(pt_model, tf_checkpoint_path, tf_inputs
...
+                unexpected_keys.add(pt_name)
+            mapped_names[pt_name] = (data, transpose)
+
+        # Some tensors could have different names but be shared
+        # We need to structures, one is to mark things as completed
+        # so we don't need to do the work again if the shared pointer
+        # was already handled.
+        # The second one, is if the shared pointer that doesn't have the shared
+        # name comes in first, we need to clear it later.
+        visited_ptr = set()
```
Comment (sgugger) [inline src/transformers/modeling_tf_pytorch_utils.py:492]:
```suggestion
        visited_pointer = set()
```

**Review by sgugger — COMMENTED** [review]
I've left a few comments on the PR but I'm extremely nervous about changing a core piece of the library like this. This function is not intended to be performant as the conversion should be done once and for all, and then the TF weights should be uploaded in the repo for every other users.

As such, I'd rather keep a version I know works than do some refactor and changes in the code.

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 2]
```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
 # TF 2.0 => PyTorch #
 #####################
 
+Shape = List[int]
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:260]: Done.

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 3]
```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
+    ptr = pt_model
+    for name in pt_name.split("."):
+        ptr = getattr(ptr, name)
+
+    array = np.array(data)
+    array = apply_transpose(array, transpose, ptr)
+    ptr.data = torch.from_numpy(array)
+
+
+def apply_transpose(array, transpose, ptr):
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:319]: You're entirely correct. `ptr` was only intended to be the temporary variable to look deep into the model (which is a moving target).

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 4]
```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
+        array = np.expand_dims(array, axis=0)
+    if list(ptr.shape) != list(array.shape):
+        squeezed_tensor_shape = [dim for dim in ptr.shape if dim != 1]
+        squeezed_array_shape = [dim for dim in array.shape if dim != 1]
+        # Making sure we don't do wrong reshape
+        if squeezed_tensor_shape == squeezed_array_shape:
+            try:
+                array = array.reshape(ptr.shape)
+            except AssertionError as e:
+                raise e
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:349]: This is a personal addition to prevent bad reshapes from ever happening (Like `zeros((2, 3)).reshape((3, 2))` which would be a logic error.

The follow up PR, completely removes the need for this code (no random squeeze/unsqueeze either)

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 5]
```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
+            except AssertionError as e:
+                e.args += (symbolic_weight.shape, array.shape)
+                raise e
+
+    try:
+        assert list(symbolic_weight.shape) == list(array.shape)
+    except AssertionError as e:
+        e.args += (symbolic_weight.shape, array.shape)
+        raise e
+    return array
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:395]: I think it is still useful (in this PR, not the follow up), since if dimensions don't match in the end we will crash here instead of later when attempting to write the data to the weight in the model.

In the follow up, `apply_transpose` don't have the reference to the tensor anymore so this check cannot exist (but operation is easily invertible)

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 6]
```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
+    elif transpose is TransposeType.CONV1D:
+        # Conv1D weight:
+        #    TF: (kernel, num_in_channel, num_out_channel)
+        # -> PT: (num_out_channel, num_in_channel, kernel)
+        array = np.transpose(array, axes=(2, 1, 0))
+    elif transpose is TransposeType.SIMPLE:
+        array = np.transpose(array)
+
+    if len(ptr.shape) < len(array.shape):
+        array = array.squeeze()
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:338]: Thanks.

File: src/transformers/modeling_tf_pytorch_utils.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:492]: Done.

File: src/transformers/modeling_tf_pytorch_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py:349]: > This is a personal addition to prevent bad reshapes from ever happening (Like zeros((2, 3)).reshape((3, 2)) which would be a logic error.

I don't think it ever happened in the tests at least but it felt like blindly reshaping could cause issues.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/modeling_tf_pytorch_utils.py]: Silently ignoring weights is absolutely not the right move, having spent several hours debugging things like that. And saying "it's going to be fixed in that other PR" for functionality at core is not an answer either.

[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py]: > Silently ignoring weights is absolutely not the right move

This is exactly what this prevents. It prevents silent errors (logic errors) and forces them to be real errors. 
The final `assert` of this function does the real erroring (Regardless of the attempt of reshaping) (I know we moved from `assert` to `if ... raise ValueError` I kept it to be closer to the original, but happy to make it `if raise ValueError`.

> "it's going to be fixed in that other PR" for functionality at core is not an answer either.

I don't think I claimed other PR solves this, I said it does not require the code anymore, meaning there's a single point of failure if tensors don't match. What I actually meant, is maybe we can follow the same path (so taking part of the other PR, and make it this one's code).

I can actually follow the same route, and make `apply_transpose` only about tensor reshaping and make validation another step as they are 2 distinct responsibility.

[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/modeling_tf_pytorch_utils.py]: At the time of my review, there was no final assert (see the [diff](https://github.com/huggingface/transformers/pull/16173/files/1ed3972bf3520f2b846f48fc8aac6342261c1dd0#diff-67993ece845388c9a9a1f342f4c82a2ed7f790f454134c39c3a63146616ea37a), so this comment is fixed now.

[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py]: My bad then ! Probably bad copy pasting from me early on.

**Comment by patrickvonplaten** [comment]
Also a bit nervous about this PR, but I see the gain we're getting with such a change! It's a good sign that all tests are passing here since we do test `from_tf=True` for all models. 

For me I think I'm generally fine with this I think. Could we maybe do the following:
- Run the changes for all pytorch and tensorflow tests cross tests (`RUN_PT_TF_CROSS_TESTS`). (No need to test the Slow tests), but we need to be sure that all cross tests are valid
- @Narsil could you specific why we need both a `apply_transpose` and a `apply_reverse` method? How is the tensor reshaping / transposing logic different from before?

**Comment by Narsil** [comment]
> Also a bit nervous about this PR, 

Would a solution like using the new code in a `try.. except` and running the exact old code if it fails make everyone feel less nervous ? (I know it would reduce my own nervousness)

> Run the changes for all pytorch and tensorflow tests cross tests (RUN_PT_TF_CROSS_TESTS). (No need to test the Slow tests), but we need to be sure that all cross tests are valid

That's already done I think. I am not sure why but I get some 5 skipped tests on brutasse, but otherwise it works, and seems to be checked by the ci already too here : 
https://app.circleci.com/pipelines/github/huggingface/transformers/36539/workflows/629c3ec0-d69f-44f2-bcec-256724806311/jobs/399239

> @Narsil could you specific why we need both a apply_transpose and a apply_reverse method? How is the tensor reshaping / transposing logic different from before?

In this PR, reshaping is very close to the same. The only thing I allowed myself to change here was to prevent logic reshape fails (Like imagine we have a `(ATTN, VOCAB_SIZE)` weight in TF, and the PT target expects `(VOCAB_SIZE, ATTN)`. Imagine the code messed up and does not declare the right `TransposeType`, then maybe we could end up doing `(ATTN, VOCAB_SIZE)reshape((VOCAB_SIZE, ATTN))` which would work technically but it's a bug IMO since tensors are different from expected. Looking at the failures when `reshape` was missing, I deduced that the code was meant to  simply handle (..., 1, ... 1 ) differences where it wouldn't necessarily be about `squeeze`/`unsqueeze` but also the position of the `1-dim` in the tensor.  Doing reshapes of that sort is perfectly safe, so I kept it as-is.

> In the follow-up (https://github.com/Narsil/transformers/pull/4) I remove this logic entirely by using new extra `TransposeType` instead. It is even scarier since TransposeType is defined purely through weight naming so if any arch doesn't have those equivalence tests, maybe we're changing tensors in an unintended way (it will crash and not be a silent error, but still).
> 
> Having `TransposeType` be invertible is IMO a necessary property to provide confidently some new functionality to save things in the `h5` instead of `pickle` with minimum changes.
> 
> ```python
> weight_name = "clip.model.conv.bias"
> tensors = torch.Tensor((x, y, z))
> 
> tf_name, transpose = transpose_type = convert_pt_weight_name_to_tf_weight_name(weight_name, tensor.shape) # Does not exist yet 
> 
> 
> # Save
> tf_tensor = apply_reverse(tensor, tranpose) 
> save_h5(tf_name, tf_tensor)
> ```
> The key thing here is that `transpose` needs only the name and `apply_reverse` would need only the original tensor, it wouldn't need to know the target.
> Reversing the name itself is a bit more challenging since it's destructive in the TF -> PT path (`prefix_to_remove`) but I haven't explored that part yet, maybe the naming is consistent enough that we can still invert the name too.
> 
> If we get there that means we could *always* save `tf_model.h5` purely from the PT model. (It does limit a little the TF naming of things, but it seems that naming is sooo consistent across the library that it's sort of already maintained/guaranteed at some point.

All of this, is to provide the least possible amount of modifications to support `h5`. If we added a new file like `pytorch_model.h5` then we can purely dictate what goes on here (but it does mean creating a split in the library serialization format, meaning people with older transformers wouldn't be able to load the new `pytorch_model.h5` files etc.. It would cause much confusion where it is used etc... Not ideal.
 
Keeping everything in already defined names means we can leverage pre-existing files and provide a safe/lower memory alternative on any model that has both PT and TF support, without modifying anything for users/models which is already quite nice !

> And **if** I get the naming thing reversible we add support to `save_pretrained(as_h5=True)` to allow saving directly in `h5` format (without needing to support the model in TF) and loading `(from_tf=True)` meaning we can also plan a smooth transition or make the choice of format a user-decision etc.. It's a big IF

**Comment by sgugger** [comment]
A thing that would make me less nervous is splitting the refactoring and the actual code changes in two different PRs so we can see more clearly the actual diff.

On the second PR that does the code change, testing the code works well for all kinds of architectures with checkpoints on the Hub would also be reassuring (it's tested in the CI on small configs but adding an integration check would be safer). We could test all checkpoints that are in the mappings with urls for instance (at the beginning of each modeling file).

**Review by LysandreJik — COMMENTED** [review]
Thanks for taking on the task!

I agree with Sylvain that splitting in two PRs would make reviews easier, and would remove some of the stress that comes from a PR touching the core serialization code.

> Would a solution like using the new code in a try.. except and running the exact old code if it fails make everyone feel less nervous ? (I know it would reduce my own nervousness)

That would absolutely help and I'd be very eager to have something like that live on `master` for a while with a good warning mentioning when it happens to test several use-cases.

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 7]
```diff
@@ -285,13 +258,146 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
 
 
+def _load(f, shapes, prefix=""):
+    """
+    Recursive internal function to load all the names of the tensors included in an h5 file
+    """
+    import h5py
+
+    for k in f.keys():
+        if isinstance(f[k], h5py._hl.dataset.Dataset):
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/modeling_tf_pytorch_utils.py]: Good catch, I looked up the concrete class , `h5py.Dataset` is better (it's an alias for `_hl.dataset.Dataset`)

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 8]
```diff
@@ -285,11 +289,164 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
...
+        )
+    except Exception as e:
+        logger.warning(
+            "We were not successful in loading the model from TF without loading the full model. This is NOT NORMAL, please warn the devs about this https://github.com/huggingface/transformers/issues/new?assignees=&labels=&template=bug-report.md&title= . Defaulting to legacy code"
+        )
+        logger.error(f"Error was {e}")
+        return legacy_load_tf2_checkpoint_in_pytorch_model(
+            pt_model, tf_checkpoint_path, tf_inputs=tf_inputs, allow_missing_keys=allow_missing_keys
+        )
+
```
Comment (Narsil) [inline src/transformers/modeling_tf_pytorch_utils.py:442]: This is the new code with `try..except` that should prevent catastophic failures.

@sgugger @LysandreJik Are you ok with the wording and how it works?

This PR is now adding code only, so you can just check the new code (and verifying `legacy_load...` is the same as the old `load_tf2_...`

**Comment by Narsil** [comment]
> On the second PR that does the code change, testing the code works well for all kinds of architectures with checkpoints on the Hub would also be reassuring

@sgugger good insight, I am finding many things which are not covered by the fast tests (mostly in the heads I think).

I am looking into it.
